### DISCUSSION TITLE
Count how many error trials in past n trials

### DISCUSSION THREAD

**Discussion by trig11** [jspsych/jsPsych#748] [Q&A] [answered]
Hi all,

I want to display an error message to participants if they make a lot of errors in a row. I think the basic layout should be an if statement, which returns true if the count of incorrect responses in the past n trials equals n. For instance, let's say I want to tell participants that they make many errors if they made three errors in a row:

```
var if_errors = {
	timeline: [errors_feedback],
	conditional_function: function(){
		var last_3_trials = jsPsych.data.get().last(3);
		var count = last_3_trials.filter({correct: false}).count();
		if (count == 3){
			return true;
			} else {
			return false;
			}
	}
}
```
I am not sure what I am missing -- but this apparently never returns true. Is there any other way to achieve this? Or what error am I making?

Note that I have different elements such as fixation and stimulus display but adding `.filter({test_part: 'response_screen'}) ` also did not help.

**Comment by jodeleeuw** [comment, marked as answer]
This looks like the right general approach to me, and I don't immediately see any bugs in the code.

If you have multiple trial parts you will need to do some kind of filter statement like the one you describe. You'll need to do that filter *before* the `.last(3)` part, or you won't get three trials with a `correct` value. The full command might look like this:

```js
var count = jsPsych.data.get().filter({test_part: 'response_screen'}).last(3).filter({correct: false}).count();
```

**Reply by trig11** [reply]
Thanks!

Your code looks right (and works if last(1) is used), but it crashes when I use anything larger than 1 into the last command, presumably because if I am in the first trial, there are no 3 last trials which can be counted in?

So I guess I would need something that checks first if there are >= 3 trials, right?

**Reply by jodeleeuw** [reply]
Ack, we should do a better job of handling that without crashing. Sorry about that!

Yes, I guess you'll have to count the number of trials first and then filter down:

```js
var test_trials = jsPsych.data.get().filter({test_part: 'response_screen'});
if(test_trials.count() >= 3){
	var count = test_trials.last(3).filter({correct: false}).count();
}
```

**Reply by trig11** [reply]
In case anyone faces the same problem, this worked for me:
```
var if_errors = {
		timeline: [errors_feedback], //should be displayed if >= 3 errors in a row
		conditional_function: function(){
		var nTrials = jsPsych.data.get().filter({test_part: 'response_screen'}).count(); //gets the number of trials available
		if (nTrials > 2) { //if there are more than 2 trials available check the correctness of the past 3
				var counter = jsPsych.data.get().filter({test_part: 'manual'}).last(3).filter({correct: false}).count(); //code from answer above
				if (counter == 3){
					return true;
					} else {
					return false;
					}
			}
			else {
				return false;
				}
			}
		}

**Reply by trig11** [reply]
> Ack, we should do a better job of handling that without crashing. Sorry about that!
> 
> Yes, I guess you'll have to count the number of trials first and then filter down:
> 
> ```js
> var test_trials = jsPsych.data.get().filter({test_part: 'response_screen'});
> if(test_trials.count() >= 3){
> 	var count = test_trials.last(3).filter({correct: false}).count();
> }
> ```

Ah, you were slightly faster :) Thanks a lot!

**Reply by jodeleeuw** [reply]
Opened an issue to get this fixed: #751
